File: NativeWrappers/Windows/SimplySaveWindows/SimplySaveWindows/MetadataStorage.cs
using SimplySaveWindows.DataTransfer;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace SimplySaveWindows
{
    public class MetadataStorage
    {
        private readonly List<NoteData> notes = new();
        private readonly List<FolderData> folders = new();

        public IReadOnlyList<NoteData> Notes { get => notes; }
        public IReadOnlyList<FolderData> Folders { get => folders; }

        public string SaveFilesDir { get; private set; }

        private const string FILE_NAME = "metadata.json";
        private string FullSaveFilePath { get => Path.Combine(SaveFilesDir, FILE_NAME); }

        public MetadataStorage(string saveFilesDir)
        {
            SaveFilesDir = saveFilesDir;

            if (File.Exists(FullSaveFilePath))
            {
                using var sr = new StreamReader(FullSaveFilePath);
                string json = sr.ReadToEnd();
                var data = JsonSerializer.Deserialize(json, typeof(MetaData)) as MetaData;
                notes.AddRange(data.Notes);
                folders.AddRange(data.Folders);
            }
        }

        public NoteData CreateNote(string title, string folderId, bool isPrivate)
        {
            int maxId = notes.Count == 0 ? -1 : notes.Max(x => Convert.ToInt32(x.Id));

            var note = new NoteData()
            {
                Id = (maxId + 1).ToString(),
                Title = title,
                FolderId = folderId,
                CreatedTimestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds(),
                EditedTimestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds(),
                IsPrivate = isPrivate
            };
            notes.Add(note);
            SaveData();
            return note;
        }

        public void NoteEdited(string id)
        {
            var note = notes.Find(x => x.Id == id);
            note.EditedTimestamp = DateTimeOffset.Now.ToUnixTimeMilliseconds();
            SaveData();
        }
        
        public void RenameNote(RenameNoteData data)
        {
            var note = notes.Find(x => x.Id == data.Id);
            note.Title = data.NewTitle;
            SaveData();
        }

        public void DeleteNote(string id)
        {
            int index = notes.FindIndex(x => x.Id == id);
            notes.RemoveAt(index);
            SaveData();
        }

        public void SetNoteFolder(string id, string newFolderId)
        {
            var note = notes.Find(x => x.Id == id);
            note.FolderId = newFolderId;
            SaveData();
        }

        private void SaveData()
        {
            if (!Directory.Exists(SaveFilesDir))
                Directory.CreateDirectory(SaveFilesDir);

            using var sw = new StreamWriter(FullSaveFilePath);
            string json = JsonSerializer.Serialize(new MetaData()
            {
                Notes = notes,
                Folders = folders
            });
            sw.WriteLine(json);
        }

        public void SetFolderData(FolderData[] newFolders)
        {
            folders.Clear();
            folders.AddRange(newFolders);
            SaveData();
        }

    }
}

File: NativeWrappers/Android/SimplySaveAndroid/app/src/main/java/com/simplysave/simplysaveandroid/MetadataStorage.kt
package com.simplysave.simplysaveandroid

import android.os.Build
import androidx.annotation.RequiresApi
import com.simplysave.simplysaveandroid.dataTransfer.FolderData
import com.simplysave.simplysaveandroid.dataTransfer.MetaData
import com.simplysave.simplysaveandroid.dataTransfer.NoteData
import com.simplysave.simplysaveandroid.dataTransfer.RenameNoteData
import kotlinx.serialization.json.Json
import kotlinx.serialization.serializer
import java.io.File
import java.time.Instant

class MetadataStorage(val saveFilesDir: String) {
    val notes: MutableList<NoteData> = mutableListOf()
    val folders: MutableList<FolderData> = mutableListOf()

    val FILE_NAME = "metadata.json"

    private fun getFullSavePath(): String {
        return File(saveFilesDir, FILE_NAME).absolutePath
    }

    init {
        val file = File(getFullSavePath())
        if (file.exists()) {
            val jsonText: String = file.readText()
            val metaData = Json.decodeFromString(serializer<MetaData>(), jsonText)
            notes.addAll(metaData.notes)
            folders.addAll(metaData.folders)
        }
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun createNote(title: String, folderId: String, isPrivate: Boolean): NoteData {
        val maxId: Int = if (notes.size == 0) -1 else notes.maxOf { note -> note.id.toInt() }

        val note = NoteData(
            id = (maxId + 1).toString(),
            title = title,
            folderId = folderId,
            createdTimestamp = Instant.now().toEpochMilli(),
            editedTimestamp = Instant.now().toEpochMilli(),
            isPrivate = isPrivate
        )
        notes.add(note)
        saveData()
        return note
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun noteEdited(id: String) {
        val note = notes.find { note -> note.id == id }
        if (note != null) {
            note.editedTimestamp = Instant.now().toEpochMilli()
        }
        saveData()
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun renameNote(data: RenameNoteData) {
        val note = notes.find { note -> note.id == data.id }
        if (note != null) {
            note.title = data.newTitle
        }
        saveData()
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun deleteNote(id: String) {
        val note = notes.find { note -> note.id == id }
        notes.remove(note)
        saveData()
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun setNoteFolder(id: String, newFolderId: String) {
        val note = notes.find { note -> note.id == id }
        if (note != null) {
            note.folderId = newFolderId
        }
        saveData()
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun saveData() {
        val dir = File(saveFilesDir)
        if (!dir.isDirectory)
            dir.mkdir()

        val jsonText = Json.encodeToString(serializer(), MetaData(notes, folders))
        val saveFile = File(getFullSavePath())
        saveFile.writeText(jsonText)
    }

    @RequiresApi(Build.VERSION_CODES.O)
    fun setFolderData(newFolders: Array<FolderData>) {
        folders.clear()
        folders.addAll(newFolders)
        saveData()
    }

}